frontend/graphwidget.cpp

```diff
@@ -10,123 +10,220 @@
 {
+    type_ = type;
     ui_->setupUi(this);
     ui_->chartView->assignChart(chart);
     ui_->binWidths->setText("Counts/ µs");
 }
 
 GraphWidget::~GraphWidget() {}
 
+QString GraphWidget::getChartRuns() { return chartRuns_; }
+QString GraphWidget::getChartDetector() { return chartDetector_; }
+QJsonArray GraphWidget::getChartData() { return chartData_; }
+
+void GraphWidget::setChartRuns(QString chartRuns) { chartRuns_ = chartRuns; }
+void GraphWidget::setChartDetector(QString chartDetector) { chartDetector_ = chartDetector; }
+void GraphWidget::setChartData(QJsonArray chartData)
+{
+    chartData_ = chartData;
+    getBinWidths();
+}
+void GraphWidget::setLabel(QString label) { ui_->statusLabel->setText(label); }
+
+void GraphWidget::getBinWidths()
+{
+    binWidths_.clear();
+    for (auto run : chartData_)
+    {
+        QVector<double> binWidths;
+        auto runArray = run.toArray();
+        for (auto i = 0; i < runArray.count() - 1; i++)
+        {
+            double binWidth = runArray.at(i + 1)[0].toDouble() - runArray.at(i)[0].toDouble();
+            binWidths.append(binWidth);
+        }
+        binWidths_.append(binWidths);
+    }
+}
+
 ChartView *GraphWidget::getChartView() { return ui_->chartView; }
 
-void GraphWidget::on_binWidths_clicked(bool checked)
+void GraphWidget::on_binWidths_toggled(bool checked)
 {
-    checked ? ui_->chartView->chart()->axes(Qt::Vertical)[0]->setTitleText("Counts/ &#181;s")
-            : ui_->chartView->chart()->axes(Qt::Vertical)[0]->setTitleText("Counts");
-    for (auto *series : ui_->chartView->chart()->series())
+    if (checked)
+    {
+        toggleOptions("binWidths");
+        ui_->chartView->chart()->axes(Qt::Vertical)[0]->setTitleText("Counts/ &#181;s");
+    }
+    else
+        ui_->chartView->chart()->axes(Qt::Vertical)[0]->setTitleText("Counts");
+    for (auto i = 0; i < ui_->chartView->chart()->series().count(); i++)
     {
-        auto xySeries = qobject_cast<QXYSeries *>(series);
+        auto xySeries = qobject_cast<QXYSeries *>(ui_->chartView->chart()->series()[i]);
         auto points = xySeries->points();
         xySeries->clear();
         if (checked)
-        {
-
-            for (auto i = 0; i < points.count() - 1; i++)
-            {
-                auto binWidth = points[i + 1].x() - points[i].x();
-                points[i].setY(points[i].y() / binWidth);
-            }
-        }
+            for (auto j = 0; j < points.count(); j++)
+                points[j].setY(points[j].y() / binWidths_[i][j]);
         else
-        {
-            for (auto i = 0; i < points.count() - 1; i++)
-            {
-                auto binWidth = points[i + 1].x() - points[i].x();
-                points[i].setY(points[i].y() * binWidth);
-            }
-        }
+            for (auto j = 0; j < points.count(); j++)
+                points[j].setY(points[j].y() * binWidths_[i][j]);
         xySeries->append(points);
```

Comment:
This, and other calls like it, are the reason for the poor performance when updating the graph. The reason is that the underlying data structure holding the points (a `QList`) is incredibly slow when appending individual points. To be honest I'm surprised that they aren't using a vector. The suggestion below increases performance dramatically, and leads to virtually no delay when changing the view toggles:
```suggestion
        xySeries->replace(points);
```